File: backend/skills/tool_list.go
package skills

import (
	"encoding/json"
	"os"
	"sort"
	"strings"

	"github.com/enough/enough/backend/config"
)

type skillItem struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	Category    string `json:"category"`
}

type listResult struct {
	Success    bool        `json:"success"`
	Skills     []skillItem `json:"skills"`
	Categories []string    `json:"categories"`
	Count      int         `json:"count"`
	Message    string      `json:"message,omitempty"`
	Hint       string      `json:"hint"`
}

func ExecuteSkillsList(argsJSON string, workDir string, cfg config.Runtime) (string, bool) {
	var args struct {
		Category string `json:"category"`
	}
	_ = json.Unmarshal([]byte(argsJSON), &args)

	skillsDir := SkillsDir()
	if _, err := os.Stat(skillsDir); os.IsNotExist(err) {
		_ = os.MkdirAll(skillsDir, 0o700)
	}

	allSkills, _ := DiscoverAllSkills(workDir, cfg)

	categoryFilter := strings.TrimSpace(args.Category)
	var filtered []Skill
	categorySet := make(map[string]bool)

	for _, sk := range allSkills {
		if categoryFilter != "" && sk.Category != categoryFilter {
			continue
		}
		filtered = append(filtered, sk)
		categorySet[sk.Category] = true
	}

	var list []skillItem
	for _, sk := range filtered {
		list = append(list, skillItem{
			Name:        sk.Name,
			Description: sk.Description,
			Category:    sk.Category,
		})
	}

	var categories []string
	for c := range categorySet {
		categories = append(categories, c)
	}
	sort.Strings(categories)

	res := listResult{
		Success:    true,
		Skills:     list,
		Categories: categories,
		Count:      len(list),
		Hint:       "Use skill_view(name) to see full content, tags, and linked files",
	}

	if len(list) == 0 {
		res.Message = "No skills found. Skills live in ~/.enough/skills/<category>/<name>/SKILL.md"
	}

	outBytes, err := json.MarshalIndent(res, "", "  ")
	if err != nil {
		return `{"success": false, "error": "json marshal error"}`, true
	}
	return string(outBytes), false
}

File: backend/skills/tool_list.ts
// PORT: backend/skills/tool_list.go

import { type runtime } from "../config/config";
import { type Skill } from "./types";
import { SkillsDir } from "./paths";
import { DiscoverAllSkills } from "./discovery";
import { Effect } from "effect";
import fs from "node:fs";

export interface skillItem {
  name: string;
  description: string;
  category: string;
}

export interface listResult {
  success: boolean;
  skills: skillItem[];
  categories: string[];
  count: number;
  message?: string;
  hint: string;
}

export function ExecuteSkillsList(
  argsJSON: string,
  workDir: string,
  cfg: runtime,
): Effect.Effect<[string, boolean], never> {
  let args = { category: "" };
  try {
    args = JSON.parse(argsJSON);
  } catch {}

  const skillsDir = SkillsDir();
  try {
    if (!fs.existsSync(skillsDir)) {
      fs.mkdirSync(skillsDir, { recursive: true, mode: 0o700 });
    }
  } catch {}

  return DiscoverAllSkills(workDir, cfg).pipe(
    Effect.match({
      onFailure: (err) => {
        return [`{"success": false, "error": ${JSON.stringify(err.message)}}`, true] as [string, boolean];
      },
      onSuccess: ([allSkills]) => {
        const categoryFilter = (args.category || "").trim();
        const filtered: Skill[] = [];
        const categorySet = new Set<string>();

        for (const sk of allSkills) {
          if (categoryFilter !== "" && sk.Category !== categoryFilter) {
            continue;
          }
          filtered.push(sk);
          categorySet.add(sk.Category);
        }

        const list = filtered.map((sk) => ({
          name: sk.Name,
          description: sk.Description,
          category: sk.Category,
        }));

        const categories = Array.from(categorySet);
        categories.sort();

        const res: listResult = {
          success: true,
          skills: list,
          categories: categories,
          count: list.length,
          hint: "Use skill_view(name) to see full content, tags, and linked files",
        };

        if (list.length === 0) {
          res.message = "No skills found. Skills live in ~/.hollow/skills/<category>/<name>/SKILL.md";
        }

        try {
          const outBytes = JSON.stringify(res, null, "  ");
          return [outBytes, false] as [string, boolean];
        } catch {
          return [`{"success": false, "error": "json marshal error"}`, true] as [string, boolean];
        }
      },
    })
  );
}

/*
PORT STATUS
source path: backend/skills/tool_list.go
source lines: 85
draft lines: 88
confidence: high
status: phase_b_compile
*/
